### ISSUE TITLE
Duplicate buttons are recorded in the navigation history

### ISSUE THREAD

**Issue by reprogrammer** [vazexqi/CodingSpectator#227]
Test `changemethodsignature.T02` presses the "Cancel" button once. But, CodingSpectator records it as being pressed twice in the "navigation-history" attribute of the resulting descriptor. The cancel button gets recorded twice whether you run the automated test or perform the steps of the test manually.

Moreover, the automated test throws the following NPE.

<pre>
org.eclipse.swt.SWTException: Failed to execute runnable (java.lang.NullPointerException)
    at org.eclipse.swt.SWT.error(SWT.java:4083)
    at org.eclipse.swt.SWT.error(SWT.java:3998)
    at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:137)
    at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:3515)
    at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3164)
    at org.eclipse.jface.window.Window.runEventLoop(Window.java:825)
    at org.eclipse.jface.window.Window.open(Window.java:801)
    at org.eclipse.ltk.ui.refactoring.RefactoringWizardOpenOperation$1.run(RefactoringWizardOpenOperation.java:186)
    at org.eclipse.swt.custom.BusyIndicator.showWhile(BusyIndicator.java:70)
    at org.eclipse.ltk.ui.refactoring.RefactoringWizardOpenOperation.run(RefactoringWizardOpenOperation.java:199)
    at org.eclipse.ltk.ui.refactoring.RefactoringWizardOpenOperation.run(RefactoringWizardOpenOperation.java:117)
    at org.eclipse.jdt.internal.ui.refactoring.actions.RefactoringStarter.activate(RefactoringStarter.java:38)
    at org.eclipse.jdt.internal.corext.refactoring.RefactoringExecutionStarter.startChangeSignatureRefactoring(RefactoringExecutionStarter.java:216)
    at org.eclipse.jdt.ui.actions.ModifyParametersAction.run(ModifyParametersAction.java:156)
    at org.eclipse.jdt.ui.actions.SelectionDispatchAction.dispatchRun(SelectionDispatchAction.java:278)
    at org.eclipse.jdt.ui.actions.SelectionDispatchAction.run(SelectionDispatchAction.java:250)
    at org.eclipse.jface.action.Action.runWithEvent(Action.java:498)
    at org.eclipse.ui.actions.RetargetAction.runWithEvent(RetargetAction.java:230)
    at org.eclipse.ui.internal.WWinPluginAction.runWithEvent(WWinPluginAction.java:234)
    at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:584)
    at org.eclipse.jface.action.ActionContributionItem.access$2(ActionContributionItem.java:501)
    at org.eclipse.jface.action.ActionContributionItem$5.handleEvent(ActionContributionItem.java:411)
    at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:84)
    at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1258)
    at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1282)
    at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1267)
    at org.eclipse.swt.widgets.Widget.notifyListeners(Widget.java:1061)
    at org.eclipse.swtbot.swt.finder.widgets.AbstractSWTBot$2.run(AbstractSWTBot.java:159)
    at org.eclipse.swtbot.swt.finder.finders.UIThreadRunnable$5.doRun(UIThreadRunnable.java:221)
    at org.eclipse.swtbot.swt.finder.finders.UIThreadRunnable$1.run(UIThreadRunnable.java:89)
    at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:35)
    at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:134)
    at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:3515)
    at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3164)
    at org.eclipse.ui.internal.Workbench.runEventLoop(Workbench.java:2640)
    at org.eclipse.ui.internal.Workbench.runUI(Workbench.java:2604)
    at org.eclipse.ui.internal.Workbench.access$4(Workbench.java:2438)
    at org.eclipse.ui.internal.Workbench$7.run(Workbench.java:671)
    at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:332)
    at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:664)
    at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:149)
    at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:115)
    at org.eclipse.swtbot.eclipse.core.UITestApplication.start(UITestApplication.java:54)
    at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
    at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:110)
    at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:79)
    at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:369)
    at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:179)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
    at java.lang.reflect.Method.invoke(Method.java:597)
    at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:620)
    at org.eclipse.equinox.launcher.Main.basicRun(Main.java:575)
    at org.eclipse.equinox.launcher.Main.run(Main.java:1408)
    at org.eclipse.equinox.launcher.Main.main(Main.java:1384)
Caused by: java.lang.NullPointerException
    at org.eclipse.ltk.internal.ui.refactoring.ErrorWizardPage.performFinish(ErrorWizardPage.java:160)
    at org.eclipse.ltk.ui.refactoring.RefactoringWizard.performFinish(RefactoringWizard.java:655)
    at org.eclipse.ltk.internal.ui.refactoring.RefactoringWizardDialog2.okPressed(RefactoringWizardDialog2.java:500)
    at org.eclipse.jface.dialogs.Dialog.buttonPressed(Dialog.java:472)
    at org.eclipse.jface.dialogs.Dialog$2.widgetSelected(Dialog.java:624)
    at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:234)
    at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:84)
    at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1258)
    at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1282)
    at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1267)
    at org.eclipse.swt.widgets.Widget.notifyListeners(Widget.java:1061)
    at org.eclipse.swtbot.swt.finder.widgets.AbstractSWTBot$2.run(AbstractSWTBot.java:159)
    at org.eclipse.swtbot.swt.finder.finders.UIThreadRunnable$5.doRun(UIThreadRunnable.java:221)
    at org.eclipse.swtbot.swt.finder.finders.UIThreadRunnable$1.run(UIThreadRunnable.java:89)
    at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:35)
    at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:134)
    ... 53 more
</pre>

**Comment by reprogrammer** [comment]
To fix this issue, I put a break point at `ChangeSignatureWizard(RefactoringWizard).addNavigationHistoryItem(NavigationHistoryItem)` and performed the steps of the test `changemethodsignature.T02` manually.

I got the following stack trace the first time the program hit the break point.

<pre>
ChangeSignatureWizard(RefactoringWizard).addNavigationHistoryItem(NavigationHistoryItem) line: 774  
RefactoringWizardDialog2.showErrorDialog(ErrorWizardPage) line: 589 
RefactoringWizardDialog2.okPressed() line: 498  
RefactoringWizardDialog2(Dialog).buttonPressed(int) line: 472   
Dialog$2.widgetSelected(SelectionEvent) line: 624   
TypedListener.handleEvent(Event) line: 234  
EventTable.sendEvent(Event) line: 84    
Button(Widget).sendEvent(Event) line: 1258  
Display.runDeferredEvents() line: 3540  
Display.readAndDispatch() line: 3161    
RefactoringWizardDialog2(Window).runEventLoop(Shell) line: 825  
RefactoringWizardDialog2(Window).open() line: 801   
RefactoringWizardOpenOperation$1.run() line: 186    
BusyIndicator.showWhile(Display, Runnable) line: 70 
</pre>


And, I got the following stack trace the second time that the program hit the break point.

<pre>
ChangeSignatureWizard(RefactoringWizard).addNavigationHistoryItem(NavigationHistoryItem) line: 775  
RefactoringWizardDialog2.cancelPressed() line: 473  
RefactoringWizardDialog2.showErrorDialog(ErrorWizardPage) line: 590 
RefactoringWizardDialog2.okPressed() line: 498  
RefactoringWizardDialog2(Dialog).buttonPressed(int) line: 472   
Dialog$2.widgetSelected(SelectionEvent) line: 624   
TypedListener.handleEvent(Event) line: 234  
EventTable.sendEvent(Event) line: 84    
Button(Widget).sendEvent(Event) line: 1258  
Display.runDeferredEvents() line: 3540  
Display.readAndDispatch() line: 3161    
RefactoringWizardDialog2(Window).runEventLoop(Shell) line: 825  
RefactoringWizardDialog2(Window).open() line: 801   
RefactoringWizardOpenOperation$1.run() line: 186    
BusyIndicator.showWhile(Display, Runnable) line: 70 
</pre>


This led me to notice the following piece of code in `org.eclipse.ltk.internal.ui.refactoring.RefactoringWizardDialog2.showErrorDialog(ErrorWizardPage)`:

```
case IDialogConstants.CANCEL_ID:
    fWizard.addNavigationHistoryItem(new NavigationHistoryItem(page.getName(), IDialogConstants.CANCEL_LABEL));
    cancelPressed();
```

Since CodingSpectator already captures the "Cancel" button in method `cancelPressed`, I removed the above call to `addNavigationHistoryItem` from `showErrorDialog` in 34b91a9f640bd4a6c7e736970e87554d30be2bd1.
